src/Type/Php/ArrayFilterFunctionReturnTypeExtension.php

```diff
@@ -78,64 +84,59 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
+			$statement = $callbackArg->stmts[0];
+			if ($statement instanceof Return_ && $statement->expr !== null) {
+				if ($mode === self::USE_ITEM) {
+					$keyVar = null;
+					$itemVar = $callbackArg->params[0]->var;
+				} elseif ($mode === self::USE_KEY) {
+					$keyVar = $callbackArg->params[0]->var;
+					$itemVar = null;
+				} elseif ($mode === self::USE_BOTH) {
+					$keyVar = $callbackArg->params[1]->var ?? null;
+					$itemVar = $callbackArg->params[0]->var;
 				}
-			} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
-				return $this->filterByTruthyValue($scope, null, $arrayArgType, $callbackArg->params[0]->var, $callbackArg->expr);
-			} elseif ($callbackArg instanceof String_) {
-				$funcName = self::createFunctionName($callbackArg->value);
-				if ($funcName === null) {
-					return new ErrorType();
-				}
-
-				$keyVar = new Variable('key');
-				$expr = new FuncCall($funcName, [new Arg($keyVar)]);
-				return $this->filterByTruthyValue($scope, null, $arrayArgType, $keyVar, $expr);
+				return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $statement->expr);
 			}
-		}
-
-		if ($flagArg instanceof ConstFetch && $flagArg->name->getParts()[0] === 'ARRAY_FILTER_USE_BOTH') {
-			if ($callbackArg instanceof Closure && count($callbackArg->stmts) === 1 && count($callbackArg->params) > 0) {
-				$statement = $callbackArg->stmts[0];
-				if ($statement instanceof Return_ && $statement->expr !== null) {
-					return $this->filterByTruthyValue($scope, $callbackArg->params[0]->var, $arrayArgType, $callbackArg->params[1]->var ?? null, $statement->expr);
-				}
-			} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
-				return $this->filterByTruthyValue($scope, $callbackArg->params[0]->var, $arrayArgType, $callbackArg->params[1]->var ?? null, $callbackArg->expr);
-			} elseif ($callbackArg instanceof String_) {
-				$funcName = self::createFunctionName($callbackArg->value);
-				if ($funcName === null) {
-					return new ErrorType();
-				}
-
-				$itemVar = new Variable('item');
-				$keyVar = new Variable('key');
-				$expr = new FuncCall($funcName, [new Arg($itemVar), new Arg($keyVar)]);
-				return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $expr);
+		} elseif ($callbackArg instanceof ArrowFunction && count($callbackArg->params) > 0) {
+			if ($mode === self::USE_ITEM) {
+				$keyVar = null;
+				$itemVar = $callbackArg->params[0]->var;
+			} elseif ($mode === self::USE_KEY) {
+				$keyVar = $callbackArg->params[0]->var;
+				$itemVar = null;
+			} elseif ($mode === self::USE_BOTH) {
+				$keyVar = $callbackArg->params[1]->var ?? null;
+				$itemVar = $callbackArg->params[0]->var;
+			}
+			return $this->filterByTruthyValue($scope, $itemVar, $arrayArgType, $keyVar, $callbackArg->expr);
+		} elseif ($callbackArg instanceof String_) {
```

Comment:
Should use `scope->getConstantStrings()` instead of raw Ast nodes

Reply:
Fixed